source/manuals/programming-languages/python/python.html.md.erb

```diff
@@ -41,6 +41,60 @@ _For more on linting (including standard settings and config) see the [linting][
 * Lint checks using [PyFlakes][]
 
 
+### Dependencies
+
+It is important to have a good strategy for specifying your project's Python
+libraries. There are two desirable characteristics:
+
+1. Reproducibility
+
+    Pin your full dependencies or you'll get unpredictability between your dev
+    environment and other environments. You want a new starter to avoid small
+    hard-to-spot problems. And you want parity between what you test locally,
+    what is tested by CI, and what you deploy, or you risk new issues appearing
+    on a live server. Additionally these things can be hard to diagnose.
+
+    Traditionally a `requirements.txt` lists the app's immediate dependencies,
+    but that leaves sub-dependencies unpinned. It makes sense to use a tool to
+    generate these, but we've noticed [minor problems with using
+    piptools/pip-compile and pipenv/Pipfile][dm-deps-commit].
```

Comment:
Given Pipfile seems to be the future way to solve this (or at least the linked commit says that) could we say here why those minor problems stop us adopting it now?

This will help future reviewers of this doc understand if we can revisit this decision.

Reply:
(Caveat: I didn't try Pipfile myself, and @allait would be better placed to comment, however...)

The linked commit explains how dependencies coming from packages installed from version control aren't being locked. This would be a pretty major problem, in fact.

Perhaps we should change the wording and just say "some problems" rather than "minor problems" :)

I didn't find an issue in github relating to that particular problem, though, so maybe we should do some further testing and get one raised.

Either way, Pipfile's readme is caveated fairly heavily with a warning "This project is under active design and development. Nothing is set in stone at this point of time." -- so I think Alexey's feeling that it isn't "ready for production" is right.

Reply:
@davbo Thanks for this. Admittedly the purpose of the link is not clear, so I've made it clear that it is the details.

I've also removed 'minor' - the severity is debatable.

Let's hope pipenv fix this issue and we can look again at adopting it. We agreed to review all this in 6 months, which has reminded me to add a line at the top of the section:

> The tools mentioned in this topic are changing quickly, so this section would benefit from review in July 2018

I've not added an expiry for the whole python document (as the GDS Way standards do) because the other manuals don't have an expiry and I don't think the choice of PEP8 and linting changes much.